### DISCUSSION TITLE
Where does Frepple execute export process?

### DISCUSSION THREAD

**Discussion by maxware0715** [frePPLe/frepple#509] [Q&A] [answered]
As title, I'm tracing export process in Frepple, which print following messages in the log file:

INFO Start step 401 'Export plan' at 09:15:23
INFO Start step export2 1 'Exporting resource plans' at 09:15:23
INFO Start step export1 1 'Exporting operationplans' at 09:15:23
INFO Finished 'Exporting resource plans' at 09:15:23
INFO Start step export2 2 'Exporting problems' at 09:15:23
INFO Finished 'Exporting problems' at 09:15:29
INFO Start step export2 3 'Exporting constraints' at 09:15:29
INFO Finished 'Exporting constraints' at 09:15:29
INFO Finished 'Exporting operationplans' at 09:17:38
INFO Start step export1 2 'Exporting operationplan materials' at 09:17:38
INFO Finished 'Exporting operationplan materials' at 09:17:40
INFO Start step export1 3 'Exporting operationplan resources' at 09:17:40
INFO Finished 'Exporting operationplan resources' at 09:17:40
INFO Start step export1 4 'Exporting demand pegging' at 09:17:40
INFO Finished 'Exporting demand pegging' at 09:17:42
INFO Start step export1 5 'Compute period of cover' at 09:17:42
INFO Finished 'Compute period of cover' at 09:17:49
INFO Finished 'Export plan' at 09:17:49

I guess the functions are defined in:
frepple/freppledb/output/commands.py

Then I get stuck and can't find where execute these classes.
Hope to know which file responses for execution.

**Comment by jdetaeye** [comment, marked as answer]
> I guess the functions are defined in:  frepple/freppledb/output/commands.py

Correct.

> Then I get stuck and can't find where execute these classes.
> Hope to know which file responses for execution.

It's structured as follows:
- The command line to run the planning algorithm is "frepple freppledb/common/commands.py".    The c++ planning engine will execute this python script with its embedded python interpretter.
- The /freppledb/common/commands.py script goes through the commands.py of all installed apps (configured in /etc/frepple/djangosettings with INSTALLED_APPS setting). The commands.py files register actions to be executed.
- Next, the planning engine executes all of the registered actions.  Some in sequence, some in parallel.

Note that the above python code execution is done by the planning engine.   This is completely separate from the django web application that is used for the user interface.    There is no communication between both processes.

**Reply by maxware0715** [reply]
Very useful. Thank you.
